### DISCUSSION TITLE
Allow the creation of requestBody

### DISCUSSION THREAD

**Discussion by savage-alex** [apideck-libraries/portman#247] [Q&A] [answered]
When variation testing, I would like to test my GET endpoints correctly reject requests that contain a requestBody.

I have a overwriteRequestBody overwrite enabled but it is not possible to create new objects / add objects to a body, only overwrite existing properties:

```json
{
                "openApiOperation": "GET::/*",
                "openApiResponse": "400",
                "variations": [
                    {
                        "name": "getWithBody",
                        "overwrites": [
                            {
                                "overwriteRequestBody": [
                                    {
                                        "key": "firstName",
                                        "value": "foo",
                                        "overwrite": true
                                    }
                                ]
                            }
                        ],
                        "tests": {
                            "contractTests": [
                                {
                                    "statusCode": {
                                        "enabled": true
                                    },
                                    "schemaValidation": {
                                        "enabled": true,
                                        "additionalProperties": false
                                    },
                                    "contentType": {
                                        "enabled": true
                                    }
                                }
                            ]
                        }
                    }
```

Is this something that would be useful to add?

(Note we can and do insert new query params in the same way to ensure unknown query params are rejected

**Comment by thim81** [comment, marked as answer]
@savage-alex 
If things work as expected, you can overwrite the request body, where by using "." as `key` this targets the root element.
As as `value` you can set an single value (like a string, number, ...) or a full object ({}) or an array ([])

```json
{
   "key": ".",
   "value": { "foo": "foo-bar-baz" }
}
```

In the Portman, you can see some examples in the [tests](https://github.com/apideck-libraries/portman/blob/main/src/application/overwrites/overwriteRequestBody.test.ts)

**Reply by savage-alex** [reply]
Wow sorry I didnt see that. :-( Shall use the tests to get more examples.

**Reply by thim81** [reply]
No need to apologise, Portman can cover already a lot of use-cases, but it is not easy to keep the documentation clear and bite-size. 
We were thinking we should create a documentation website for it, but havent gotten around to it.

Plus this discussions section, could help others as well :-)

**Reply by savage-alex** [reply]
Progress! 
The good, When using overwriteRequestBody with a PUT operation (already has a body), I can see the new value!

The Bad, but its taking one character at a time:
![image](https://user-images.githubusercontent.com/25430683/150307379-1b30e3c3-a60d-4602-9387-74f614f89fee.png)

When I flip the openAPIOperation to target GET operations, nothing is seen at all, my theory here is that Postman is in this mode:
![image](https://user-images.githubusercontent.com/25430683/150307605-ac991df5-4ca6-4a15-855d-c68f4bd0d6e0.png)
Unless we set the Body to raw it doesnt seem to be in postman (its not in the generated collection)

Files to reproduce if you wish:
Simple portman config:
[portman-config.adv.body.overwrite.json.txt](https://github.com/apideck-libraries/portman/files/7903859/portman-config.adv.body.overwrite.json.txt)

OpenAPI with get and put operation:
[AdvancedComputerSoft-demo-advanced-car-inventory-1.1.2-swagger.yaml.txt](https://github.com/apideck-libraries/portman/files/7903864/AdvancedComputerSoft-demo-advanced-car-inventory-1.1.2-swagger.yaml.txt)

**Reply by nicklloyd** [reply]
Interesting behaviour.... And if you set the value as an object instead of a string?

**Reply by savage-alex** [reply]
Shall try (with some help from @AliKhalili :-) ), Interestingly changing overwrite to false = 
![image](https://user-images.githubusercontent.com/25430683/150311267-5c31be2b-06b5-42df-b728-aa21a734fad5.png)

**Reply by thim81** [reply]
I delivered a bad example, the object should be an "object" (without the "") instead of string.

**Reply by savage-alex** [reply]
This overwriteRequestBody will add the following key at root level
```
 {
                                "overwriteRequestBody": [
                                    {
                                        "key": "foo",
                                        "value": "foo-bar-baz' ",
                                        "overwrite": true
                                    }
                                ]
                            }
```
Perfect test to see how a PUT operation will behave with an unknown property.
![image](https://user-images.githubusercontent.com/25430683/150351636-d72f578b-0fdc-4f61-8054-e89cd23926d5.png)

Sadly the issue with GET remains where it is not created at all.

**Reply by AliKhalili** [reply]
@savage-alex I think adding request body for those operations which do not have request body is a legitimate variation test.
Portman ignores this scenario, @thim81, Could you please tell us is there any specific reason behind that?
It would be my pleasure to add this new feature with a PR.

**Reply by thim81** [reply]
The idea behind Portman is to build tests based on the defined properties in OpenAPI for the Happy path and fiddle with the properties that are defined for the variations. This to reduce the repetitive work of defining the same request/responses everywhere (docs, tests, ...). OpenAPI is our single source for everything, hence the term "spec-driven development".

So if it doesn't exist in OpenAPI, Portman will have trouble to deduce test from it and build a Postman request. Example: If there is no response body, then Portman cannot generate tests to verify the expected, documented response. If there is no request body defined, it is difficult to create one, since it would miss plenty of request context.

**Reply by savage-alex** [reply]
Hi @thim81, thanks for the considered response on this. It makes total sense. 
We saw this case as an opportunity to automate the repetitive task for our testers to ensure compliance with the spec: 
* If the spec says there can be no body, did the API correctly reject one that was sent. 
* We do this currently with a special postman operation that we hijack bodies into via a script. 
* The variation or fuzzing of injecting things like headers and query prams works great (I have built a case for an unknown query param variation which works brilliantly). This was the next case we planned to try and automate as its something we have fallen fowl of before (its one from my playbook thats made other peoples APIs fall over.)

Perhaps we misunderstood the true purpose of portman and are trying to mod it to the point it is a contact tester for all our positive and negative scenarios (we have header tests, query tests, body tests that I am trying to automate). For me contract compliance is about doing what you promised in the OpenAPI which exists for both doing what I said I would and only that. If its not in the definition, I am going to reject it (I just want to make sure we are doing that correctly)

Thanks
